### PR TITLE
[WIP][SPARK-1486][MLlib] Multi Model Training with Gradient Descent

### PR DESCRIPTION
**Pull request by brkyvz**
**Note: This is still a work in progress**

This is the first of the pull requests to support multi-model training in MLlib. It batches examples and trains multiple models with different regularization parameters and step sizes all at once using Matrix-Matrix multiplication. It uses Native BLAS when the data matrix is dense, and uses sparse matrices as much as possible for both better memory utilization and performance (I will post performance results in the comments).

This is a HUGE Pull Request, therefore I'm posting this now. It is not finished, docs need to be updated, code can be somewhat cleaned up for ease of understanding. I'm posting this now so that users can comment and make suggestions along the way.

Most of the PR consists of adding additional Local Matrix operations for the calculation of gradients and losses.

### CODE REVIEW

**Comment by brkyvz** [comment]
closing this PR as a lot of functionality has changed
